### PR TITLE
Enhancement: Merge Better Login Reports Into Core.

### PR DESCRIPTION
**Pull request by andrewlimaza**
Merge "Better Logins Report Add On" into PMPro core (overwritten default functionality).

Added in admin notice on reports page if Add On is still activated when on PMPro 2.0, this will have a quick link to deactivate the Better Login Reports Add On but not delete the plugin. (Only deactivates).

### CODE REVIEW

File: adminpages/reports/login.php [hunk 1]
```diff
@@ -3,60 +3,73 @@
...
 	$now = current_time('timestamp');
-	$visits = get_option("pmpro_visits", array("today"=>0, "thisday"=>date_i18n("Y-m-d", $now), "alltime"=>0, "month"=>0, "thismonth"=>date_i18n("n", $now)));
-	$views = get_option("pmpro_views", array("today"=>0, "thisday"=>date_i18n("Y-m-d", $now), "alltime"=>0, "month"=>0, "thismonth"=>date_i18n("n", $now)));
-	$logins = get_option("pmpro_logins", array("today"=>0, "thisday"=>date_i18n("Y-m-d", $now), "alltime"=>0, "month"=>0, "thismonth"=>date_i18n("n", $now)));
+	
+	$visits = pmpro_login_report_all_values('visits');
+	$views = pmpro_login_report_all_values('views');
+	$logins = pmpro_login_report_all_values('logins');
 ?>
-<span id="pmpro_report_login" class="pmpro_report-holder">
```
Comment (ideadude) [inline adminpages/reports/login.php]: This class is needed for v2.0 updates to reports we've done.

File: adminpages/reports/login.php [hunk 2]
```diff
@@ -3,60 +3,73 @@
...
+		</tr>
+		<tr>
+			<th scope="row"><?php _e('All Time','paid-memberships-pro'); ?></th>
 			<td><?php echo number_format_i18n($visits['alltime']); ?></td>
 			<td><?php echo number_format_i18n($views['alltime']);?></td>
 			<td><?php echo number_format_i18n($logins['alltime']); ?></td>
 		</tr>
 	</tbody>
-	</table>
-	<?php if ( function_exists( 'pmpro_report_login_page' ) ) { ?>
```
Comment (ideadude) [inline adminpages/reports/login.php]: Double check this.

File: adminpages/reports/login.php [hunk 3]
```diff
@@ -66,32 +79,27 @@ function pmpro_report_login_widget()
 function pmpro_report_login_page()
 {
 	global $wpdb;
-	$now = current_time('timestamp');
-
+	
 	//vars
 	if(!empty($_REQUEST['s']))
-		$s = sanitize_text_field($_REQUEST['s']);
+		$s = $_REQUEST['s'];
```
Comment (ideadude) [inline adminpages/reports/login.php]: sanitize_text_field($_REQUEST['s']);

File: adminpages/reports/login.php [hunk 4]
```diff
@@ -66,32 +79,27 @@ function pmpro_report_login_widget()
...
-
+	
 	//vars
 	if(!empty($_REQUEST['s']))
-		$s = sanitize_text_field($_REQUEST['s']);
+		$s = $_REQUEST['s'];
 	else
 		$s = "";
-
-	if(!empty($_REQUEST['l'])) {
```
Comment (ideadude) [inline adminpages/reports/login.php]: Use this instead of code from BLR

File: adminpages/reports/login.php [hunk 5]
```diff
@@ -66,32 +79,27 @@ function pmpro_report_login_widget()
...
+	else
 		$l = "";
-	}
 ?>
-	<form id="posts-filter" method="get" action="">
-	<h1>
-		<?php _e('Visits, Views, and Logins Report', 'paid-memberships-pro' );?>
-	</h1>
+	<form id="posts-filter" method="get" action="">	
+	<h2>
```
Comment (ideadude) [inline adminpages/reports/login.php]: Should be H1 now.

File: adminpages/reports/login.php [hunk 6]
```diff
@@ -66,32 +79,27 @@ function pmpro_report_login_widget()
...
+		<?php _e('Visits, Views, and Logins Report', 'paid-memberships-pro');?>
+	</h2>		
 	<ul class="subsubsub">
-		<li>
-			<?php _e('Show', 'paid-memberships-pro' )?> <select name="l" onchange="jQuery('#posts-filter').submit();">
-				<option value="" <?php if(!$l) { ?>selected="selected"<?php } ?>><?php _e('All Users', 'paid-memberships-pro' )?></option>
-				<option value="all" <?php if($l == "all") { ?>selected="selected"<?php } ?>><?php _e('All Levels', 'paid-memberships-pro' )?></option>
+		<li>			
+			<?php _ex('Show', 'Dropdown label, e.g. Show All Users', 'paid-memberships-pro')?> <select name="l" onchange="jQuery('#posts-filter').submit();">
```
Comment (ideadude) [inline adminpages/reports/login.php]: Switch to just _e(). _ex() confuses our pot generator.

File: adminpages/reports/login.php [hunk 7]
```diff
@@ -101,97 +109,104 @@ function pmpro_report_login_page()
...
 				?>
-			</select>
+			</select>			
 		</li>
 	</ul>
 	<p class="search-box">
-		<label class="hidden" for="post-search-input"><?php _e('Search', 'paid-memberships-pro' )?> <?php if(empty($l)) echo "Users"; else echo "Members";?>:</label>
-		<input type="hidden" name="page" value="pmpro-reports" />
-		<input type="hidden" name="report" value="login" />
+		<label class="hidden" for="post-search-input"><?php _ex('Search', 'Search form label', 'paid-memberships-pro')?> <?php if(empty($l)) echo "Users"; else echo "Members";?>:</label>
```
Comment (ideadude) [inline adminpages/reports/login.php]: Switch back to _e()

File: adminpages/reports/login.php [hunk 8]
```diff
@@ -101,97 +109,104 @@ function pmpro_report_login_page()
...
+		<input type="hidden" name="page" value="pmpro-reports" />		
+		<input type="hidden" name="report" value="better_login" />
```
Comment (ideadude) [inline adminpages/reports/login.php]: This needs to be just "login"

File: adminpages/reports/login.php [hunk 9]
```diff
@@ -101,97 +109,104 @@ function pmpro_report_login_page()
...
 		{
 			$sqlQuery = "SELECT SQL_CALC_FOUND_ROWS u.ID, u.user_login, u.user_email, UNIX_TIMESTAMP(u.user_registered) as joindate, mu.membership_id, mu.initial_payment, mu.billing_amount, mu.cycle_period, mu.cycle_number, mu.billing_limit, mu.trial_amount, mu.trial_limit, UNIX_TIMESTAMP(mu.startdate) as startdate, UNIX_TIMESTAMP(mu.enddate) as enddate, m.name as membership FROM $wpdb->users u LEFT JOIN $wpdb->usermeta um ON u.ID = um.user_id LEFT JOIN $wpdb->pmpro_memberships_users mu ON u.ID = mu.user_id AND mu.status = 'active' LEFT JOIN $wpdb->pmpro_membership_levels m ON mu.membership_id = m.id WHERE (u.user_login LIKE '%" . esc_sql($s) . "%' OR u.user_email LIKE '%" . esc_sql($s) . "%' OR um.meta_value LIKE '%" . esc_sql($s) . "%') ";
-
+		
 			if($l == "all")
 				$sqlQuery .= " AND mu.status = 'active' AND mu.membership_id > 0 ";
 			elseif($l)
-				$sqlQuery .= " AND mu.membership_id = '" . esc_sql($l) . "' ";
-
+				$sqlQuery .= " AND mu.membership_id = '" . $l . "' ";
```
Comment (ideadude) [inline adminpages/reports/login.php]: esc_sql( $l );

File: adminpages/reports/login.php [hunk 10]
```diff
@@ -101,97 +109,104 @@ function pmpro_report_login_page()
...
 		{
 			$sqlQuery = "SELECT SQL_CALC_FOUND_ROWS u.ID, u.user_login, u.user_email, UNIX_TIMESTAMP(u.user_registered) as joindate, mu.membership_id, mu.initial_payment, mu.billing_amount, mu.cycle_period, mu.cycle_number, mu.billing_limit, mu.trial_amount, mu.trial_limit, UNIX_TIMESTAMP(mu.startdate) as startdate, UNIX_TIMESTAMP(mu.enddate) as enddate, m.name as membership FROM $wpdb->users u LEFT JOIN $wpdb->pmpro_memberships_users mu ON u.ID = mu.user_id AND mu.status = 'active' LEFT JOIN $wpdb->pmpro_membership_levels m ON mu.membership_id = m.id";
 			$sqlQuery .= " WHERE 1=1 ";
-
+			
 			if($l == "all")
 				$sqlQuery .= " AND mu.membership_id > 0  AND mu.status = 'active' ";
 			elseif($l)
-				$sqlQuery .= " AND mu.membership_id = '" . esc_sql($l) . "' ";
+				$sqlQuery .= " AND mu.membership_id = '" . $l . "' ";
```
Comment (ideadude) [inline adminpages/reports/login.php]: esc_sql( $l );

File: adminpages/reports/login.php [hunk 11]
```diff
@@ -202,43 +217,49 @@ function pmpro_report_login_page()
...
 								</strong>
-							</td>
+							</td>										
 							<td>
 								<?php echo $theuser->display_name;?>
 							</td>
-							<td><?php echo $auser->membership?></td>
-							<td><?php echo date_i18n("m/d/Y", strtotime($theuser->user_registered, current_time("timestamp")))?></td>
+							<td><?php echo $auser->membership?></td>												
+							<td><?php echo date("m/d/Y", strtotime($theuser->user_registered, current_time("timestamp")))?></td>
```
Comment (ideadude) [inline adminpages/reports/login.php]: date_i18n( ... )

File: adminpages/reports/login.php [hunk 12]
```diff
@@ -202,43 +217,49 @@ function pmpro_report_login_page()
...
 							<td>
-								<?php
-									if($auser->enddate)
-										echo date_i18n(get_option('date_format'), $auser->enddate);
+								<?php 									
+									if($auser->enddate) 
+										echo date(get_option('date_format'), $auser->enddate);
```
Comment (ideadude) [inline adminpages/reports/login.php]: date_i18n( ... )

File: adminpages/reports/login.php [hunk 13]
```diff
@@ -252,194 +273,265 @@ function pmpro_report_login_page()
 /*
 	Other code required for your reports. This file is loaded every time WP loads with PMPro enabled.
 */
-
-//track visits
-function pmpro_report_login_wp_visits()
+//get values for a user
+function pmpro_login_report_get_values_for_user($type, $user_id = NULL)
 {
-	//don't track admin
```
Comment (ideadude) [inline adminpages/reports/login.php]: we prob want this back if it's not still in there.

File: adminpages/reports/login.php [hunk 14]
```diff
@@ -252,194 +273,265 @@ function pmpro_report_login_page()
...
+		//get values
+		$values = pmpro_login_report_get_values_for_user($type, $user_id);
+
+		if($values !== false)
 		{
-			$views['month'] = 1;
-			$views['thismonth'] = $thismonth;
+			//track for user
+			$values['last'] = date(get_option("date_format"), $now);
+			$values['alltime']++;
```
Comment (ideadude) [inline adminpages/reports/login.php]: Change all of these ++ to something like:

`$values['alltime'] = $values['alltime'] + 1;`

There is a note RE a PHP bug fix on these in the old version.

File: includes/updates.php [hunk 15]
```diff
@@ -135,3 +135,26 @@ function pmpro_updates_notice_complete() {
 </div>
 <?php
 }
+
```
Comment (ideadude) [inline includes/updates.php]: Cool for now. I might move this into a separate file when we bundle all of the 2.0 updates stuff together.

File: adminpages/reports/login.php (hunk 13 above)
Comment (andrewlimaza) [reply adminpages/reports/login.php]: This is the function that would replace the pmpro_report_login_wp_visits - https://github.com/strangerstudios/pmpro-better-logins-report/blob/master/reports/better-logins.php#L390

Would this be okay to replace or should we keep this function for tracking visits?

Comment (ideadude) [reply adminpages/reports/login.php]: This function pmpro_report_login_wp_visits will become a wrapper like in BLR:

```
/track visits
function pmpro_report_better_login_wp_visits()
{
	pmpro_report_login_track_values("visits");	
}
add_action("wp", "pmpro_report_better_login_wp_visits");
```

The pmproblr_trackValues renamed to pmpro_report_login_track_values and use code from BLR

**Pull request merged**